File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.js
// BFS
function solution(begin, target, words) {
    let ans = 0
    
    // 포함이 안되어 있으면 return 0
    const firstCheck = words.filter(x => x === target)
    if (firstCheck.length === 0) return 0
    
    const q = [[begin, 0]]
    
    while (q.length !== 0) {
        let [newWord, cnt] = q.shift() // 배열에서 첫 번째 요소 제거
        
        if (newWord === target) return cnt // target에 도착하면 return
        
        words.forEach((word, index) => {
            let diff = 0
            
            for (let i=0; i < word.length; ++i) {
                if (word[i] !== newWord[i]) diff += 1
            }
            
            if (diff === 1) q.push([word, cnt+1])
        })
        
    }
    
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
answer = []


def dfs(begin, target, words, count, visited):
    if begin == target:
        global answer
        answer.append(count)
        
    for index in range(len(words)):
        onediff = 0
        if visited[index] == 0:
            
            for i in range(len(begin)):
                if begin[i] != words[index][i]:
                    onediff += 1

            if onediff == 1:
                visited[index] = 1
                dfs(words[index], target, words, count + 1, visited)
                visited[index] = 0


def solution(begin, target, words):
    # 변환 불가인 경우
    if target not in words:
        return 0
    
    visited = [0] * len(words)
    dfs(begin, target, words, 0, visited)
    return min(answer)